[Intelligence Effects] Document markers are not hidden during intelligence effect animations

rr-codes opened the pull request:
#### 7d00282bd5ca61f15a047bc7fe67b6f6e6ab0032
<pre>
[Intelligence Effects] Document markers are not hidden during intelligence effect animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=281780">https://bugs.webkit.org/show_bug.cgi?id=281780</a>
<a href="https://rdar.apple.com/138209992">rdar://138209992</a>

Reviewed by Abrar Rahman Protyasha.

Currently, when intelligence effects are active (such as during a Writing Tools rewrite operation), transparent
content document markers are added to the affected range with the intention of hiding all content within that range.
However, they were only hiding the associated text, and were not hiding other document markers.

To fix, the transparent content markers are combined with the other markers such that the transparent content markers
take precedence during any overlapping ranges with other markers during the subdivision operation.

Also adds testing support infrastructure to facilitate adding markers from JavaScript.

* LayoutTests/editing/spelling/transparent-content-marker-expected.html: Added.
* LayoutTests/editing/spelling/transparent-content-marker.html: Added.
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectionStartHasMarkerFor const):
(WebCore::Editor::selectionStartSetMarkerFor):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarkers):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setMarkerFor):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/285455@main">https://commits.webkit.org/285455@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d365e1b0be679d3e255e2a55fc35bed9e0d278de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72730 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52155 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25529 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74845 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59960 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23764 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75797 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/47164 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62595 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20061 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/22293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20420 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/16976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/78594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/17024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62603 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78594 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16048 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47953 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2740 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/49020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/50315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/48765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/Editor.h:
```diff
@@ -518,6 +518,7 @@ class Editor final : public CanMakeCheckedPtr<Editor> {
     WEBCORE_EXPORT void replaceSelectionWithFragment(DocumentFragment&, SelectReplacement, SmartReplace, MatchStyle, EditAction = EditAction::Insert, MailBlockquoteHandling = MailBlockquoteHandling::RespectBlockquote);
     WEBCORE_EXPORT void replaceSelectionWithText(const String&, SelectReplacement, SmartReplace, EditAction = EditAction::Insert);
     WEBCORE_EXPORT bool selectionStartHasMarkerFor(DocumentMarker::Type, int from, int length) const;
+    WEBCORE_EXPORT void selectionStartSetMarkerFor(DocumentMarker::Type, int from, int length, const String&);
```
I wish the name could better reflect its purpose (for testing only)? Any ideas?

rr-codes replied:
ya I don’t have any ideas, I just copy and pasted the getter name essentially 😅

whsieh replied:
Right, a `…ForTesting` prefix would help clarify that this method is (under normal circumstances) reachable from tests only. You could also use `RELEASE_ASSERT_NOT_REACHED_WITH_MESSAGE` instead of `RELEASE_ASSERT_NOT_REACHED` in that method, and then you don't need the comment anymore.

aprotyas commented on Source/WebCore/rendering/TextBoxPainter.cpp:
```diff
@@ -1055,8 +1055,25 @@ void TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarkers()
+
+    // Ensure the transparent content marked texts go first in the vector, so that they take precedence over
+    // the other marked texts when being subdivided so that they do not get painted.
+    Vector<MarkedText> allMarkedTexts;
+    allMarkedTexts.appendVector(transparentContentMarkedTexts);
+    allMarkedTexts.appendVector(markedTexts);
```
Instead of making a new vector, consider:

```
markedTexts.insertVector(0, transparentContentMarkedTexts);
```

rr-codes replied:
I considered this, I did it this way for less mutability of variables which makes things easier to read imo. I can change it if you feel strongly though?

aprotyas replied:
Yes I see your point. I suggested this to avoid an extra vector copy, but I don't have a strong opinion.

rr-codes committed:
[Intelligence Effects] Document markers are not hidden during intelligence effect animations
https://bugs.webkit.org/show_bug.cgi?id=281780
rdar://138209992

Reviewed by Abrar Rahman Protyasha.

Currently, when intelligence effects are active (such as during a Writing Tools rewrite operation), transparent
content document markers are added to the affected range with the intention of hiding all content within that range.
However, they were only hiding the associated text, and were not hiding other document markers.

To fix, the transparent content markers are combined with the other markers such that the transparent content markers
take precedence during any overlapping ranges with other markers during the subdivision operation.

Also adds testing support infrastructure to facilitate adding markers from JavaScript.

* LayoutTests/editing/spelling/transparent-content-marker-expected.html: Added.
* LayoutTests/editing/spelling/transparent-content-marker.html: Added.
* Source/WebCore/editing/Editor.cpp:
(WebCore::Editor::selectionStartHasMarkerFor const):
(WebCore::Editor::selectionStartSetMarkerFor):
* Source/WebCore/editing/Editor.h:
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarkers):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setMarkerFor):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/285455@main
